libevm/VMOpt.cpp

```diff
@@ -0,0 +1,154 @@
+/*
+	This file is part of cpp-ethereum.
+
+	cpp-ethereum is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	cpp-ethereum is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+
+#include "VM.h"
+#include "VMConfig.h"
+#include <libethereum/ExtVM.h>
+using namespace std;
+using namespace dev;
+using namespace dev::eth;
+
+void VM::reportStackUse()
+{
+	static intptr_t p = 0;
+	intptr_t q = intptr_t(&q);
+	if (p)
+		cerr << "STACK: " << p << " - " << q << " = " << (p - q) << endl;
+	p = q;
+}
+
+std::array<InstructionMetric, 256> VM::c_metrics;
+void VM::initMetrics()
+{
+	static bool done=false;
+	if (!done)
+	{
+		for (unsigned i = 0; i < 256; ++i)
+		{
+			InstructionInfo inst = instructionInfo((Instruction)i);
+			c_metrics[i].gasPriceTier = inst.gasPriceTier;
+			c_metrics[i].args = inst.args;
+			c_metrics[i].ret = inst.ret;
+		}
+	}
+	done = true;
+}
+
+//
+// init interpreter on entry
+//
+
+void VM::initEntry()
+{
+	m_bounce = &VM::interpretCases;	
+	m_code_vector = m_ext->code;
+	m_code = m_code_vector.data();
+	
+	initMetrics();	
+	
+	optimize();
+}
+
+int VM::pool_constant(const u256& _con)
+{
+	int i = 0, n = m_pool_vector.size();
+	while (i < n)
```

Comment:
if you mean `if (mPoolSpace[i] == _con)` then both are u256.   But I can reduce the number of comparisons by using a map to find the unique constants, then copy it to a vector when done.  Another possibility is to wait until encountering a PUSHn to put the constant in the pool and replace the PUSHn with PUSHC.  Once it's working I can look at that.